conda_lock/models/lock_spec.py

```diff
@@ -24,23 +28,141 @@ class _BaseDependency(StrictModel):
     def sorted_extras(cls, v: List[str]) -> List[str]:
         return sorted(v)
 
+    def _merge_base(self, other: _BaseDependency) -> _BaseDependency:
+        if other is None:
+            return self
+        if (
+            self.name != other.name
+            or self.manager != other.manager
+            or self.category != other.category
+        ):
+            raise ValueError(
+                "Cannot merge incompatible dependencies: {self} != {other}"
+            )
+        return _BaseDependency(
+            name=self.name,
+            manager=self.manager,
+            category=self.category,
+            extras=list(set(self.extras + other.extras)),
+        )
+
 
 class VersionedDependency(_BaseDependency):
     version: str
     build: Optional[str] = None
     conda_channel: Optional[str] = None
 
+    @staticmethod
+    def _merge_matchspecs(
+        matchspec1: Optional[str],
+        matchspec2: Optional[str],
+        combine_constraints: bool = True,
+    ) -> Optional[str]:
+        if matchspec1 == matchspec2:
+            return matchspec1
+        if matchspec1 is None or matchspec1 == "":
+            return matchspec2
+        if matchspec2 is None or matchspec2 == "":
+            return matchspec1
+        if fnmatchcase(matchspec1, matchspec2):
+            return matchspec1
+        if fnmatchcase(matchspec2, matchspec1):
+            return matchspec2
+        if not combine_constraints:
+            raise ValueError(
+                f"Found incompatible constraint {matchspec1}, {matchspec2}"
+            )
+        return f"{matchspec1},{matchspec2}"
+
+    def merge(self, other: Optional[VersionedDependency]) -> VersionedDependency:
+        if other is None:
+            return self
+
+        if (
+            self.conda_channel is not None
+            and other.conda_channel is not None
+            and self.conda_channel != other.conda_channel
+        ):
+            raise ValueError(
+                f"VersionedDependency has two different conda_channels:\n{self}\n{other}"
+            )
+        merged_base = self._merge_base(other)
+        try:
+            build = self._merge_matchspecs(
+                self.build, other.build, combine_constraints=False
+            )
```

Comment:
> Also I think you can get rid of the surrounding try/except block: just raise the exception directly in _merge_build.

The surrounding try/except block adds the package names to the exception, which is convenient for the user to figure out which package has conflicting builds. Otherwise the error message would only contain the build ids